src/BrainPass/BrainPass.sol

```diff
@@ -0,0 +1,284 @@
+        return baseTokenURI;
+    }
+
+    function setBaseURI(string memory tokenURI) internal {
+        baseTokenURI = tokenURI;
+    }
+
+    /// @notice Add a new Pass Type
+    /// @param passId and others are the details needed for a passType
+    function addPassType(
+        uint256 passId,
+        uint256 pricePerDay,
+        string memory tokenURI,
+        string memory name,
+        uint256 maxTokens,
+        uint256 discount
+    ) public onlyOwner {
+        require(bytes(tokenURI).length > 0, "Invalid token URI");
+        require(maxTokens > 0, "Invalid max tokens");
+
+        passTypes[passId] = PassType(
+            passId,
+            name,
+            pricePerDay,
+            tokenURI,
+            maxTokens,
+            discount,
+            0
+        );
+
+        emit NewPassAdded(passId, name, maxTokens, pricePerDay);
+    }
+
+    /// @notice Mint and NFT of a particular passtype
+    /// @param passId The id of the passtype to mint
+    function mintNFT(
+        uint256 passId,
+        uint256 startTimestamp,
+        uint256 endTimestamp
+    ) public {
+        require(
+            addressToPassId[msg.sender][passId] != true,
+            "Max NFTs per address reached"
+        );
+
+        PassType storage passType = passTypes[passId];
+
+        require(passType.maxTokens != 0, "Pass type not found");
+
+        require(
+            passType.lastTokenIdMinted.add(1) <= passType.maxTokens,
+            "Max supply reached"
+        );
+
+        uint256 price = calculatePrice(passId, startTimestamp, endTimestamp);
+
+        if (IERC20(IqToken).balanceOf(msg.sender) < price)
+            revert UserBalanceNotEnough();
+
+        bool success = IERC20(IqToken).transferFrom(msg.sender, owner, 1e18);
+        if (!success) revert MintingPaymentFailed();
+
+        uint256 tokenId = passType.lastTokenIdMinted + 1;
+        setBaseURI(passType.tokenURI);
+        _safeMint(msg.sender, tokenId);
+
+        addressToPassId[msg.sender][passId] = true;
+
+        UserPassItem memory purchase = UserPassItem(
+            tokenId,
+            passId,
+            startTimestamp,
+            endTimestamp
+        );
+        addressToNFTPass[msg.sender][tokenId] = purchase;
+        passType.lastTokenIdMinted = tokenId;
+
+        emit BrainPassBought(
+            msg.sender,
+            tokenId,
+            passId,
+            startTimestamp,
+            endTimestamp
+        );
+    }
+
+    /// @notice Calculate the price of an Nft
+    /// @param startTimestamp and endTimestamp are used to calc the price to be paid
+    function calculatePrice(
+        uint256 passId,
+        uint256 startTimestamp,
+        uint256 endTimestamp
+    ) public view returns (uint256) {
+        PassType memory passType = passTypes[passId];
+        //duration in days
+        uint256 duration = endTimestamp.sub(startTimestamp);
+        uint256 totalPrice = duration.mul(passType.pricePerDay);
```

Comment:
if start and end are timestamp, then you do the diff and multiply per price perday? timestamp are seconds not days